### PR TITLE
V3 Version

### PR DESCRIPTION
**Pull request by fabrice-quadrata**
## Description

V3 includes the following:
- update in schema in how we store attribute [instead of storing a list of attributes, we store a mapping (key => Attribute) with the key being a keccak256(userAddress, attrType, issuerAddress) or keccak256(userDID, attrType, issuerAddress)]
- this reduces gas cost of minting by 50%
- no changes in gas cost for query
- Added benefit: if an issuer is deleted, all attributes issued by this issuer won't be queryable
- remove unused functions like `latestEpoch` or `hasAttributesByIssuer`, which will get replaced by attributeMetadata

## Checklist
1. [x] Unit Testing?
2. [ ] Backward Compatible?
3. [x] Upgrade Safe (storage)?
4. [ ] Updated Confluence?
5. [ ] Updated Gitbook?

### CODE REVIEW

File: contracts/QuadPassport.sol [hunk 1]
```diff
@@ -356,6 +404,51 @@ contract QuadPassport is IQuadPassport, UUPSUpgradeable, PausableUpgradeable, Qu
         emit SetPendingGovernance(pendingGovernance);
     }
 
+    /// @dev Allow users to grab all the metadata for a given attribute(s)
+    /// @param _account address of user
+    /// @param _attributes attributes to get respective non-value data from
+    /// @return attributeNames list of attribute names encoded as keccack256("AML") for example
+    /// @return issuers list of issuers for the attribute[i]
+    /// @return issuedAts list of epochs for the attribute[i]
+    function attributeMetadata(
```
Comment (fabrice-quadrata) [inline contracts/QuadPassport.sol:413]: Thanks @teddy-clapp ,

Can we add back also your original implementation on top?
attributeExists(address _account, bytes32[] memory _attributes) public view returns(bytes32[] attributeNames, bool[] exist)

Comment (teddy-clapp) [reply contracts/QuadPassport.sol:410]: We might not have enough space for both. I’ll finish getting the tests passing and see where are at with “optimizer runs: 0”

Comment (teddy-clapp) [reply contracts/QuadPassport.sol:427]: I set runs to 0, we cannot add many more functions to QuadReader now. one big one, or a few small ones

Comment (fabrice-quadrata) [reply contracts/QuadPassport.sol:427]: ok, let's add them to QuadReader then?

File: contracts/QuadPassport.sol [hunk 2]
```diff
@@ -356,6 +401,82 @@ contract QuadPassport is IQuadPassport, UUPSUpgradeable, PausableUpgradeable, Qu
         emit SetPendingGovernance(pendingGovernance);
     }
 
+    function _attributeLength(
```
Comment (fabrice-quadrata) [inline contracts/QuadPassport.sol:404]: add docstring

File: contracts/QuadPassport.sol [hunk 3]
```diff
@@ -356,6 +401,82 @@ contract QuadPassport is IQuadPassport, UUPSUpgradeable, PausableUpgradeable, Qu
...
+                }
+            }
+        }
+    }
+
+    /// @dev Allow users to grab all existences for a given attribute(s)
+    /// @param _account address of user
+    /// @param _attributes attributes to get respective bool data from
+    /// @return existences list of bools for the attribute[i]
+    function attributeExists(
```
Comment (fabrice-quadrata) [inline contracts/QuadPassport.sol:459]: attributesExist

File: contracts/QuadPassport.sol [hunk 4]
```diff
@@ -435,4 +556,47 @@ contract QuadPassport is IQuadPassport, UUPSUpgradeable, PausableUpgradeable, Qu
...
+        uint256 eligibleAttributesLength = governance.getEligibleAttributesLength();
+        for(uint256 i = 0; i < eligibleAttributesLength; i++) {
+            bytes32 eligibleAttribute = governance.eligibleAttributesArray(i);
+            // loop over all accounts
+            for(uint256 j = 0; j < _accounts.length; j++) {
+                address account = _accounts[j];
+                bytes32 attrKeyv1;
+                if(governance.eligibleAttributesByDID(eligibleAttribute)) {
+                    Attribute[] memory did = _attributes[keccak256(abi.encode(account, eligibleAttribute))];
+                    require(did.length > 0, "INVALID_DID");
```
Comment (fabrice-quadrata) [inline contracts/QuadPassport.sol:579]: and did.value != bytes32(0)

File: contracts/QuadPassport.sol [hunk 5]
```diff
@@ -435,4 +556,47 @@ contract QuadPassport is IQuadPassport, UUPSUpgradeable, PausableUpgradeable, Qu
...
+                }
+                Attribute[] memory attributesV1 = _attributes[attrKeyv1];
+                // loop over attributes and write to _attributesv2
+                for(uint256 k = 0; k < attributesV1.length; k++) {
+                    Attribute memory attributeV1 = attributesV1[k];
+                    bytes32 attKey = attributeKey(account, eligibleAttribute, attributeV1.issuer);
+                    _attributesv2[attKey] = IQuadPassportStore.Attribute({
+                        value: attributeV1.value,
+                        epoch: attributeV1.epoch,
+                        issuer: attributeV1.issuer
```
Comment (fabrice-quadrata) [inline contracts/QuadPassport.sol:592]: address(0)

File: contracts/QuadPassport.sol [hunk 6]
```diff
@@ -435,4 +556,47 @@ contract QuadPassport is IQuadPassport, UUPSUpgradeable, PausableUpgradeable, Qu
...
+                    });
+                }
+
+                delete _attributes[attrKeyv1];
```
Comment (fabrice-quadrata) [inline contracts/QuadPassport.sol:596]: you sure it deletes all items?

File: contracts/QuadPassport.sol [hunk 7]
```diff
@@ -435,4 +556,47 @@ contract QuadPassport is IQuadPassport, UUPSUpgradeable, PausableUpgradeable, Qu
...
+    /// @param _accounts list of accounts to migrate
+    function migrateAttributes(address[] calldata _accounts) external {
+        // check sender has governance role
+        require(
+            IAccessControlUpgradeable(address(governance)).hasRole(GOVERNANCE_ROLE, _msgSender()),
+            "INVALID_ADMIN"
+        );
+
+        // loop over all attributes by did/account
+        uint256 eligibleAttributesLength = governance.getEligibleAttributesLength();
```
Comment (fabrice-quadrata) [inline contracts/QuadPassport.sol:570]: Sanity check: Can you double check that DID is listed after AML today?

File: contracts/interfaces/IQuadPassport.sol [hunk 8]
```diff
@@ -30,6 +30,12 @@ interface IQuadPassport is IQuadSoulbound {
         bytes calldata _sigIssuer
     ) external payable;
 
+    function attributeMetadata(
```
Comment (fabrice-quadrata) [inline contracts/interfaces/IQuadPassport.sol:33]: missing attributesExist

File: contracts/interfaces/IQuadPassport.sol [hunk 9]
```diff
@@ -38,6 +44,8 @@ interface IQuadPassport is IQuadSoulbound {
 
     function acceptGovernance() external;
 
+    function attribute(address _account, bytes32 _attribute) external view returns (IQuadPassportStore.Attribute memory);
```
Comment (fabrice-quadrata) [inline contracts/interfaces/IQuadPassport.sol:47]: maybe add attrKey

File: contracts/QuadGovernance.sol [hunk 10]
```diff
@@ -209,6 +210,8 @@ contract QuadGovernance is IQuadGovernance, AccessControlUpgradeable, UUPSUpgrad
 
                 revokeRole(ISSUER_ROLE, _issuer);
 
+                _deletedIssuers.push(_issuer);
```
Comment (fabrice-quadrata) [inline contracts/QuadGovernance.sol:213]: This works, but if you add an already deleted issuers, you have to remove it from _deletedIssuers

File: test/QuadPassport/test_migrate_to_v3_storage.ts [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+            });
+
+            console.log("migrated chunk", chunk)
+        }
+
+
+
+        // ensure all attributes in all accounts are equal to premigrationAttributes
+        for (let account of chunks[0]) {
+            const attributes = await quadReader.connect(signers[0]).callStatic.getAttributesBulk(account, eligibleAttributes, {
```
Comment (fabrice-quadrata) [inline test/QuadPassport/test_migrate_to_v3_storage.ts:149]: can we make sure attributes.length >0

File: contracts/QuadPassport.sol (hunk 7 above)
Comment (teddy-clapp) [reply contracts/QuadPassport.sol:570]: good check, this wasn't the case today. doing some work to fix it

File: contracts/QuadGovernance.sol (hunk 10 above)
Comment (fabrice-quadrata) [reply contracts/QuadGovernance.sol:213]: Either fix in this PR, or create a ticket.

File: contracts/QuadPassport.sol (hunk 2 above)
Comment (teddy-clapp) [reply contracts/QuadPassport.sol:408]: fixed

File: contracts/QuadPassport.sol (hunk 4 above)
Comment (teddy-clapp) [reply contracts/QuadPassport.sol:585]: fixed

File: contracts/QuadPassport.sol (hunk 5 above)
Comment (teddy-clapp) [reply contracts/QuadPassport.sol]: fixed

File: contracts/QuadPassport.sol (hunk 7 above)
Comment (teddy-clapp) [reply contracts/QuadPassport.sol]: this opened up a rabbit hole. it has been fixed and resolved, there were two bugs that resulted in AML in not getting migrated. both have been patched and tests now include AML in the checks

File: contracts/interfaces/IQuadPassport.sol (hunk 8 above)
Comment (teddy-clapp) [reply contracts/interfaces/IQuadPassport.sol:33]: it's there hoss

File: contracts/interfaces/IQuadPassport.sol (hunk 9 above)
Comment (teddy-clapp) [reply contracts/interfaces/IQuadPassport.sol:47]: should work, let's see what CI says

File: contracts/QuadGovernance.sol (hunk 10 above)
Comment (teddy-clapp) [reply contracts/QuadGovernance.sol:213]: fixed

File: test/QuadPassport/test_migrate_to_v3_storage.ts (hunk 11 above)
Comment (teddy-clapp) [reply test/QuadPassport/test_migrate_to_v3_storage.ts:160]: checks out:

`expect(attributes.length).to.be.equal(eligibleAttributes.length);`

fixed

File: contracts/QuadPassport.sol (hunk 6 above)
Comment (teddy-clapp) [reply contracts/QuadPassport.sol:600]: yes sir

**Review by teddy-clapp — APPROVED** [review]

**Review by faze-quadrata — APPROVED** [review]

**Pull request merged**